src/System.Linq/src/System/Linq/SelectMany.cs

```diff
@@ -133,5 +123,122 @@ public static partial class Enumerable
                 }
             }
         }
+
+        private sealed class SelectManySingleSelectorIterator<TSource, TResult> : Iterator<TResult>, IIListProvider<TResult>
+        {
+            private readonly IEnumerable<TSource> _source;
+            private readonly Func<TSource, IEnumerable<TResult>> _selector;
+            private IEnumerator<TSource> _sourceEnumerator;
+            private IEnumerator<TResult> _subEnumerator;
+
+            internal SelectManySingleSelectorIterator(IEnumerable<TSource> source, Func<TSource, IEnumerable<TResult>> selector)
+            {
+                Debug.Assert(source != null);
+                Debug.Assert(selector != null);
+
+                _source = source;
+                _selector = selector;
+            }
+
+            public override Iterator<TResult> Clone()
+            {
+                return new SelectManySingleSelectorIterator<TSource, TResult>(_source, _selector);
+            }
+
+            public override void Dispose()
+            {
+                if (_sourceEnumerator != null)
+                {
+                    _sourceEnumerator.Dispose();
+                    _sourceEnumerator = null;
+                }
+
```

Comment:
`_subEnumerator` may not be null here, and need to be disposed too, if the enumeration is jumped out of.